### PR TITLE
pass on hydration top level field arguments to isFieldForbidden

### PR DESCRIPTION
**Pull request by temaEmelyan**
not sure that's the way forward, hence the (SPIKE).

We need that to correctly calculate whether a field is allowed or not in the implementation of `isFIeldForbidden`.

### CODE REVIEW

File: src/main/java/graphql/nadel/hooks/HydrationArguments.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import graphql.language.Argument;
+import graphql.schema.GraphQLArgument;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+@PublicApi
+public class HydrationArguments {
```
Comment (andimarek) [inline src/main/java/graphql/nadel/hooks/HydrationArguments.java:13]: I could imagine we make this more convenient by providing get(name), getValue(name) like methods

Comment (temaEmelyan) [reply src/main/java/graphql/nadel/hooks/HydrationArguments.java:15]: added these methods

File: src/main/java/graphql/nadel/hooks/HydrationArguments.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        this.variables = Objects.requireNonNull(builder.variables);
+        this.hydrationRootGqlArguments = builder.hydrationRootGqlArguments == null ? Collections.emptyList() : builder.hydrationRootGqlArguments;
+        this.hydrationRootArguments = builder.hydrationRootArguments == null ? Collections.emptyList() : builder.hydrationRootArguments;
+    }
+
+    public List<GraphQLArgument> getHydrationRootGqlArguments() {
+        return hydrationRootGqlArguments;
+    }
+
+    public List<Argument> getHydrationRootArguments() {
```
Comment (andimarek) [inline src/main/java/graphql/nadel/hooks/HydrationArguments.java:31]: I would recommend to name this `AstArguments` to make it more clear.

File: src/test/groovy/graphql/nadel/RemovedFieldsTest.groovy [hunk 3]
```diff
@@ -849,4 +850,373 @@ class RemovedFieldsTest extends StrategyTestHelper {
...
+        type User {
+            id: ID
+            restricted: String
+        }
+        """)
+        def query = 'query issueQ($productId: ID) {issue(id: "ISSUE-1") { id author(productId: $productId) { id restricted }} }'
+        given:
+        def hooks = new ServiceExecutionHooks() {
+            @Override
+            CompletableFuture<Optional<GraphQLError>> isFieldForbidden(NormalizedQueryField normalizedField, HydrationArguments executionParameters, Object userSuppliedContext) {
```
Comment (andimarek) [inline src/test/groovy/graphql/nadel/RemovedFieldsTest.groovy:1087]: `HydrationArguments` is named `executionParameters`? that looks weird.

Comment (andimarek) [reply src/test/groovy/graphql/nadel/RemovedFieldsTest.groovy:1087]: also: I would not expect that you access the variables directly: I think you are only interested in argument values. So there should be a `getArgumentValue` method on `HydrationArguments`

**Review by andimarek — APPROVED** [review]
small things ... see comments. In general good

File: engine/src/main/java/graphql/nadel/engine/execution/AsyncIsFieldForbidden.java [hunk 4]
```diff
@@ -22,13 +22,20 @@
 
     private final Map<NormalizedQueryField, GraphQLError> fieldsToErrors = new ConcurrentHashMap<>();
     private final ServiceExecutionHooks serviceExecutionHooks;
-    private final ExecutionContext executionContext;
     private final NadelContext nadelContext;
+    private final HydrationArguments hydrationArguments;
+    private final Map<String, Object> variables;
 
-    public AsyncIsFieldForbidden(ServiceExecutionHooks serviceExecutionHooks, ExecutionContext executionContext, NadelContext nadelContext) {
+    public AsyncIsFieldForbidden(
```
Comment (andimarek) [inline engine/src/main/java/graphql/nadel/engine/execution/AsyncIsFieldForbidden.java:29]: lets use the new `nonNullable/nullable` annotations here to indicate what could be null or not...

File: api/src/main/java/graphql/nadel/hooks/HydrationArguments.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        } else {
+            this.hydrationRootGqlArguments = hydrationRootGqlArguments;
+            this.hydrationRootAstArguments = hydrationRootAstArguments;
+        }
+    }
+
+    public static HydrationArguments empty() {
+        return new HydrationArguments(Collections.emptyList(), Collections.emptyList());
+    }
+
```
Comment (andimarek) [inline api/src/main/java/graphql/nadel/hooks/HydrationArguments.java:31]: these are underlying schema `GraphQLArgument`, right?

Comment (temaEmelyan) [reply api/src/main/java/graphql/nadel/hooks/HydrationArguments.java:31]: no, it's overall, see this test https://github.com/atlassian-labs/nadel/pull/194/files#diff-4d0ae59bc268469697cf0a858dd7d447c88b139fd3dbf646ccb2592a82e64ecbR1185

File: engine/src/main/java/graphql/nadel/engine/execution/AsyncIsFieldForbidden.java (hunk 4 above)
Comment (temaEmelyan) [reply engine/src/main/java/graphql/nadel/engine/execution/AsyncIsFieldForbidden.java:35]: added `org.jetbrains.annotations.NotNull` to contructor args, fields, and the public method

**Review by andimarek — APPROVED** [review]

**Pull request merged**